src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -137,41 +138,104 @@ class GcsRpcClient {
+      gcs_is_down_ = true;
+    } else {
+      gcs_is_down_ = false;
+    }
+
     job_info_grpc_client_ =
-        std::make_unique<GrpcClient<JobInfoGcsService>>(channel, client_call_manager);
+        std::make_unique<GrpcClient<JobInfoGcsService>>(channel_, client_call_manager);
     actor_info_grpc_client_ =
-        std::make_unique<GrpcClient<ActorInfoGcsService>>(channel, client_call_manager);
+        std::make_unique<GrpcClient<ActorInfoGcsService>>(channel_, client_call_manager);
     node_info_grpc_client_ =
-        std::make_unique<GrpcClient<NodeInfoGcsService>>(channel, client_call_manager);
+        std::make_unique<GrpcClient<NodeInfoGcsService>>(channel_, client_call_manager);
     node_resource_info_grpc_client_ =
-        std::make_unique<GrpcClient<NodeResourceInfoGcsService>>(channel,
+        std::make_unique<GrpcClient<NodeResourceInfoGcsService>>(channel_,
                                                                  client_call_manager);
     heartbeat_info_grpc_client_ = std::make_unique<GrpcClient<HeartbeatInfoGcsService>>(
-        channel, client_call_manager);
+        channel_, client_call_manager);
     stats_grpc_client_ =
-        std::make_unique<GrpcClient<StatsGcsService>>(channel, client_call_manager);
+        std::make_unique<GrpcClient<StatsGcsService>>(channel_, client_call_manager);
     worker_info_grpc_client_ =
-        std::make_unique<GrpcClient<WorkerInfoGcsService>>(channel, client_call_manager);
+        std::make_unique<GrpcClient<WorkerInfoGcsService>>(channel_, client_call_manager);
     placement_group_info_grpc_client_ =
-        std::make_unique<GrpcClient<PlacementGroupInfoGcsService>>(channel,
+        std::make_unique<GrpcClient<PlacementGroupInfoGcsService>>(channel_,
                                                                    client_call_manager);
     internal_kv_grpc_client_ =
-        std::make_unique<GrpcClient<InternalKVGcsService>>(channel, client_call_manager);
+        std::make_unique<GrpcClient<InternalKVGcsService>>(channel_, client_call_manager);
     internal_pubsub_grpc_client_ = std::make_unique<GrpcClient<InternalPubSubGcsService>>(
-        channel, client_call_manager);
+        channel_, client_call_manager);
+
+    // Setup monitor for gRPC channel status.
+    // TODO(iycheng): Push this into ClientCallManager with CQ by using async call.
+    periodical_runner_->RunFnPeriodically(
+        [this] {
+          if (shutdown_) {
+            return;
+          }
+          auto status = channel_->GetState(true);
+          // https://grpc.github.io/grpc/core/connectivity__state_8h_source.html
+          RAY_LOG(DEBUG) << "GCS channel status: " << status;
+          switch (status) {
+          case GRPC_CHANNEL_TRANSIENT_FAILURE:
+          case GRPC_CHANNEL_CONNECTING:
+            if (!gcs_is_down_) {
+              gcs_is_down_ = true;
+            } else {
+              RAY_CHECK(absl::ToInt64Seconds(absl::Now() - gcs_last_alive_time_) <
```

Comment:
should we make some check to ensure gcs_rpc_server_reconnect_timeout_s always greater than the gcs reconnect retry timeout?

Reply:
I feel it's not necessary here given that channel reconnection is a dynamic thing: 100ms for the initial one and 2s for the maximum one with exponential backoff.

Any concerns here for this?